### PR TITLE
Implement RFC #4873 improvements to JavaScript FFI

### PR DESCRIPTION
**Pull request by mjendrusch**
Adds types and macros to address RFC #4873. The types added, `JsObject` and `JsAssoc` come with both bracket accessors, and experimental dot accessors, as well as a `.()` operator.

### CODE REVIEW

**Comment by mjendrusch** [comment]
Tests failing is probably due to node.js, as the version employed by travis seems to be v0.10.36, and node.js does not support the JavaScript spread operator `...` used in the implementation of `.()`. The spread operator is supported without additional flags from node.js v5.12.0 onwards.
Should I change the implementation to use `Function.prototype.apply` to support more versions of node.js, or is ok using node.js v5.12.0 and newer (current stable is v7.4.0)?

**Comment by vegansk** [comment]
Please use ES5 standard to avoid incompatibility problems

**Comment by mjendrusch** [comment]
Done. With the last commit everything should be compatible to ECMAScript 3rd edition, the earliest IE supported with the features used should be IE6.

File: lib/js/jsffi.nim [hunk 1]
```diff
@@ -0,0 +1,479 @@
...
+##  # This uses an experimental ``.()`` operator for ``JsObject``, to emit
+##  # JavaScript calls, when no corresponding proc exists for ``JsObject``.
+##  proc main =
+##    jq(document).ready(proc() =
+##      console.log("Hello JavaScript!")
+##    )
+##
+## JsFFI also provides macros to facilitate fully statically typed interaction
+## with Javascript. To this end, one often has to write types, where every field
+## is marked with ``exportc`` to avoid mangling. For types like React attributes
```
Comment (Araq) [inline lib/js/jsffi.nim]: This needs to be fixed in the JS code generator, not hacked around!

File: lib/js/jsffi.nim [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+  when K is string:
+    yield $k
+  else:
+    yield k
+  {.emit: "}".}
+
+# Literal generation
+
+macro lit*(xs: varargs[untyped]): auto =
+  ## Takes a ``typedesc`` as its first argument, and a series of expressions of
```
Comment (Araq) [inline lib/js/jsffi.nim]: So make it take a ``typedesc`` as its first argument!

File: lib/js/jsffi.nim (hunk 1 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: Would the fix needed be to forego mangling of fields and field accesses as done in `genFieldAddr` and `genFieldAccess` in module `jsgen` (changing the default mangling behaviour for only the js target)? If so, I'll try to fix this in a separate pull request and remove `pragmaTypeSection` from this one, if not, could you please point me in the right direction?

Comment (Araq) [inline lib/js/jsffi.nim]: > If so, I'll try to fix this in a separate pull request and remove pragmaTypeSection from this one.

Yes, sounds right.

File: lib/js/jsffi.nim (hunk 2 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: Now takes a `typedesc` as its first argument and an `untyped` as its second.
The rationale behind this is to stay as close as possible to the previous syntax, which is not achievable using `typedesc` and `varargs[untyped]`, as an expression `Typ.lit(a = 11, b = 22)` is parsed as a call with named arguments, and not a `varargs[untyped]`, which is why a macro with those argument types could not be called.

File: lib/js/jsffi.nim [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+    setImpl[K, V](obj, field, val)
+
+proc `==`*(x, y: JsRoot): bool =
+  ## Compare two JsObjects or JsAssocs. Be careful though, as this is comparison
+  ## like in JavaScript, so if your JsObjects are in fact JavaScript Objects,
+  ## and not strings or numbers, this is a *comparison of references*.
+  compareImpl(x, y)
+
+{. experimental .}
+proc `.`*(obj: JsObject, field: cstring): JsObject =
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: This should better use dot-notation in the resulting codegen and be a template ;)

File: lib/js/jsffi.nim [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+  ## Example:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##  let obj = newJsObject()
+  ##  obj.a = 20
+  ##  console.log(obj.a) # puts 20 onto the console.
+  getAsImpl[cstring, JsObject](obj, field)
+
+proc `.=`*[T](obj: JsObject, field: cstring, value: T) =
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: ditto

File: lib/js/jsffi.nim [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+  {.emit: "  if (!`assoc`.hasOwnProperty(`k`)) continue;".}
+  when K is string:
+    yield $k
+  else:
+    yield k
+  {.emit: "}".}
+
+# Literal generation
+
+macro lit*(typ: typedesc, xs: untyped = []): auto =
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: Would not it be more consistent to use the associative array literals?
Here's a working example: https://gist.github.com/yglukhov/dfac1b81cbd714c5b8da8cc4ba57e4d8

File: lib/js/jsffi.nim [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+
+proc setImpl[K, V](obj: JsRoot, field: K, val: V)
+  {.importcpp: "#[#] = #" .}
+
+proc getAsImpl[K, V](obj: JsRoot, field: K): V
+  {. importcpp: "#[#]" .}
+
+proc callFieldAsImpl[K, V](obj: JsRoot, field: K, args: varargs[JsObject]): V =
+  when V isnot void:
+    {. emit: [result, "=", obj, "[", field, "].apply(", obj, ", ", args, ");"] .}
```
Comment (vegansk) [inline lib/js/jsffi.nim]: Did I miss something and interpolation in ``emit`` is deprecated now?

File: lib/js/jsffi.nim [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+##
+
+when not defined(js) and not defined(nimdoc) and not defined(nimsuggest):
+  {.fatal: "Module jsFFI is designed to be used with the JavaScript backend.".}
+
+import macros
+
+const
+  setImpl = "#[#] = #"
+  setImplJsString = "#[toJSStr(#)] = #"
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: `toJSStr` is not guaranteed to be present in the resulting code. For it to appear Nim has to know that it is used. What you should do instead is just change proc argument to `cstring`, and importcpp would be just `#[#] = #`.

File: lib/js/jsffi.nim [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+type
+  JsRoot* = ref object of RootObj
+    ## Root type of both JsObject and JsAssoc
+  JsObject* = ref object of JsRoot
+    ## Dynamically typed wrapper around a JavaScript object.
+  JsAssoc*[K, V] = ref object of JsRoot
+    ## Statically typed wrapper around a JavaScript object.
+
+# Import variables:
+template magicVar*(name, typ: untyped): untyped =
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: TBH I'm not sure in the usefullness of this template:
```nim
var myVar {.importc, nodecl.}: MyType
# vs
magicVar(myVar, MyType)
```
is not a huge difference. But the first statement is explicit, understandable and searchable.

File: lib/js/jsffi.nim [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+  var `name` {. importc, nodecl, inject .}: `typ`
+
+# New
+proc newJsObject*: JsObject {. importcpp: "{@}" .}
+  ## Creates a new empty JsObject
+proc newJsAssoc*[K, V]: JsAssoc[K, V] = {. importcpp: "{@}" .}
+  ## Creates a new empty JsAssoc with key type `K` and value type `V`.
+
+# Checks
+proc hasOwnProperty*(x: JsObject, prop: string): bool
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: The bug is still there. s/string/cstring.

File: lib/js/jsffi.nim [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+  {. importcpp: setImplJsString .}
+  ## Set the value of a property of name `field` in a JsAssoc `obj` to `v`.
+
+proc `==`*(x, y: JsRoot): bool {. importcpp: "(# === #)" .}
+  ## Compare two JsObjects or JsAssocs. Be careful though, as this is comparison
+  ## like in JavaScript, so if your JsObjects are in fact JavaScript Objects,
+  ## and not strings or numbers, this is a *comparison of references*.
+
+{. experimental .}
+proc `.`*(obj: JsObject, field: cstring): JsObject {. importcpp: getImpl .}
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: Not that critical, but I would try to leverage `static[cstring]` powers for the resulting js code to use dot-notation ;)

File: lib/js/jsffi.nim [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+  ##
+  ## Example:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##  let obj = newJsObject()
+  ##  obj.a = 20
+  ##  console.log(obj.a) # puts 20 onto the console.
+
+proc `.=`*[T](obj: JsObject, field: cstring, value: T) {. importcpp: setImpl .}
```
Comment (yglukhov) [inline lib/js/jsffi.nim]: ditto

File: lib/js/jsffi.nim (hunk 6 above)
Comment (yglukhov) [inline lib/js/jsffi.nim]: @vegansk: http://nim-lang.org/news/e029_version_0_16_0.html, look for "emit"

File: lib/js/jsffi.nim (hunk 9 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: Done. Sorry, I missed that.

File: lib/js/jsffi.nim (hunk 10 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: I'm a bit reluctant to use dot-notation here, as in nim there are a variety of valid field names (`if`, `for`), which cause dot-notation to break for old versions of JavaScript (like version 1.5, [which nim is supposed to emit](http://nim-lang.org/docs/backends.html#backends-the-javascript-target)), or even all versions (`==`, and other kinds of operator-like identifiers).
So even if those field names are somewhat unorthodox, I think they should be supported.

File: lib/js/jsffi.nim (hunk 8 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: You're right, it is trivial and now removed :)

**Comment by yglukhov** [comment]
As a side-note. I can't see a use-case where `{.importc.}` on a var stmt would not require `{.nodecl.}`. Maybe we should consider `{.nodecl.}` behavior automatically for such cases in the codegen.

File: lib/js/jsffi.nim (hunk 10 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: Never mind, I'll try to implement it in analogy to `jsgen`'s `mangleName`.

**Comment by Araq** [comment]
@yglukhov I agree.

File: lib/js/jsffi.nim (hunk 10 above)
Comment (mjendrusch) [inline lib/js/jsffi.nim]: Now all dot-accessors generate dot-notation JavaScript.

**Comment by Araq** [comment]
@dom96 Should this be part of the stdlib?
@yglukhov Everything fine now?

**Comment by yglukhov** [comment]
Also instead of using `apply` you can do some nim tricks to avoid that (see https://github.com/nim-lang/Nim/blob/devel/lib/js/jsconsole.nim). That surely can wait unless you have mood to try it in this PR =)

**Comment by mjendrusch** [comment]
Now, we don't generate any `apply` anywhere, anymore.
To be honest, I also like the associative-array-notation better, though bound to a type, an instance of which it generates:

```nim
var obj = JsObject{ a: 22, b: "foo" }

# or

type MyType = object
  a: int
  b: cstring

var myObj = MyType{ a: 10 }
```

**Comment by yglukhov** [comment]
🥇 Supercool! Thanks a lot for your work!

**Comment by dom96** [comment]
This is absolutely awesome! Thank you so much for writing this. I'm definitely happy to have this in the stdlib. Please verify if you can that the docgen for this module is correct (I think the ``.. code-block`` might need some more indentation), also add it [here](https://github.com/nim-lang/Nim/blob/devel/web/website.ini#L36).

**Comment by mjendrusch** [comment]
I'm glad to help :) I will probably be using this module quite heavily in the future.
The documentation renders just fine, I should probably add a link to it in `lib.rst` as well.
Just out of curiosity, what is the system behind the grouping of modules in `website.ini`, and should I just add the module to line 36 as indicated?

**Comment by dom96** [comment]
> Just out of curiosity, what is the system behind the grouping of modules in website.ini, and should I just add the module to line 36 as indicated?

As far as I know there is no system. Just place it wherever you wish :)

**Comment by mjendrusch** [comment]
Wonderful, it was just a tiny bit confusing to me. It should be added everywhere it needs to be to show up in the documentation now.

**Comment by Araq** [comment]
> Just out of curiosity, what is the system behind the grouping of modules in website.ini, and should I just add the module to line 36 as indicated?

Also 'website.ini' needs to die. We need blacklisting of .nim files instead of whitelisting for the docgen.

File: lib/js/jsffi.nim [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+    "void", "while", "with", "yield", "enum", "implements", "interface",
+    "let", "package", "private", "protected", "public", "static", "await",
+    "abstract", "boolean", "byte", "char", "double", "final", "float", "goto",
+    "int", "long", "native", "short", "synchronized", "throws", "transient",
+    "volatile", "null", "true", "false"]
+  case name
+  of reservedWords: return false
+  else: discard
+  for chr in name:
+    if chr notin {'A'..'Z','a'..'z','_'}:
```
Comment (Araq) [inline lib/js/jsffi.nim]: '$', '0'..'9'

Comment (mjendrusch) [inline lib/js/jsffi.nim]: Sorry, I missed that.

**Comment by andreaferretti** [comment]
This has already lead to [removing 130 lines out of 300](https://github.com/andreaferretti/react.nim/commit/e393dbb3cdf1c55651dbd0cfa1360428341bea44#diff-0d318cdf205c001ab99b114797efbcea) in my react wrapper, and there is probably room for more simplification :-)

**Pull request merged**